src/provider/model/Storage.ts

```diff
@@ -1,13 +1,95 @@
 import localforage from 'localforage';
+import {injectable} from 'inversify';
 
-export const settingsStorage = localforage.createInstance({
-    driver: localforage.INDEXEDDB,
-    name: 'notifications',
-    storeName: 'settings'
-});
-
-export const notificationStorage = localforage.createInstance({
-    driver: localforage.INDEXEDDB,
-    name: 'notifications',
-    storeName: 'notifications'
-});
+import {AsyncInit} from '../controller/AsyncInit';
+
+/**
+ * List of available storages.
+ */
+export const enum StorageMap {
+    NOTIFICATIONS = 'notifications',
+    SETTINGS = 'settings'
+}
+
+@injectable()
+export class Storage extends AsyncInit {
+    /**
+     * This value should be updated any time changes are made to the notification shape.
+     * Must remain in number format.  LocalForage does not support semver.
+     */
+    public static readonly DATABASE_VERSION: number = 1.0;
+    private _storages: Map<string, LocalForage>;
+
+    constructor() {
+        super();
+        this._storages = new Map<string, LocalForage>();
+
+        [StorageMap.NOTIFICATIONS, StorageMap.SETTINGS].forEach(storeName => {
+            this.add({
+                driver: localforage.INDEXEDDB,
+                version: Storage.DATABASE_VERSION,
+                name: 'notifications',
+                storeName
+            });
+        });
+    }
+
+    protected async init(): Promise<void> {
+        await Promise.all([...this._storages.values()].map(storage => storage.ready()));
+
+        const settingsStorage = this.get(StorageMap.SETTINGS);
+
+        const storedVersion = await settingsStorage.getItem<number>('dbVersion');
+        const dbVersion = settingsStorage.config().version!;
+
+        if (!storedVersion) {
+            await this.initializeDatabase();
+        } else if (storedVersion < dbVersion) {
+            console.log(`Upgrading database from version ${storedVersion} to ${dbVersion}`);
+            await this.upgradeDatabase();
+        }
+    }
+
+    private add(options: LocalForageOptions): void {
+        const instance = localforage.createInstance(options);
+        this._storages.set(options.storeName!, instance);
+    }
+
+    private async upgradeDatabase(): Promise<void> {
+        const settingsStorage = this.get(StorageMap.SETTINGS);
+
+        const dbVersion = settingsStorage.config().version!;
+        const storedVersion = await settingsStorage.getItem<number>('dbVersion');
+
+        // Example code.  Fill in actual upgrade paths here.
```

Comment:
Can we do this upgrade atomically, since there's two different stores? We'd need to both upgrade the notifications store, and bump the number is the settings atomically.

Is this something we need to worry about?

Reply:
I don't think we need to worry about it too much.  The settings I imagine should cover the version for both storages, and if we need to update one or the other we can still do that in the steps.  I tried to leave it pretty open for future arrangement when things are actually upgraded.